File: test/integration/dispatchers.py
"""Dispatcher ownership across real shared-library unloads."""

from pathlib import Path


def lifecycle(s):
    fixture = s.artifact("test-dispatchers.so")
    assert fixture.is_file(), "build integration-fixtures before the dispatcher suite"

    def load(path):
        s.ctl("plugin", "load", str(path))
        assert str(path) in Path(f"/proc/{s.compositor_pid}/maps").read_text()

    def unload(path):
        s.ctl("plugin", "unload", str(path))

    def reply(name, expected):
        result = s.run("hyprctl", "dispatch", name)
        assert expected in result, (name, result)
        assert s.compositor.poll() is None, "dispatcher invocation crashed the compositor"
        assert s.data("monitors"), "compositor IPC stopped responding"

    def unmapped(path):
        # A dependency or GNU-unique symbol must not accidentally keep the
        # provider mapped and turn an unload regression into a false pass.
        assert str(path) not in Path(f"/proc/{s.compositor_pid}/maps").read_text()

    s.close()
    unload(s.plugin)
    load(fixture)
    load(s.plugin)
    reply("hyprspace-test:ping", "fixture:original")
    reply("hyprspace-test:legacy", "ok")
    for name in ("raw", "closure"):
        reply("hyprspace-test:" + name, "fixture:raw")
    unload(fixture)
    unmapped(fixture)
    unload(s.plugin)
    reply("hyprspace-test:ping", "Invalid dispatcher")
    for name in ("legacy", "raw", "closure"):
        reply("hyprspace-test:" + name, "Invalid dispatcher")
    s.check("B loads before hyprspace: unloading B then hyprspace never resurrects B's callback")

    load(s.plugin)
    load(fixture)
    s.ctl("dispatch", "hyprspace-test:replace", "replacement")
    reply("submap", "fixture:replacement")
    unload(s.plugin)
    unmapped(s.plugin)
    reply("submap", "fixture:replacement")
    reply("hyprspace-test:ping", "fixture:original")
    s.check("a handler installed after hyprspace survives hyprspace unloading")
    unload(fixture)
    unmapped(fixture)
    reply("submap", "Invalid dispatcher")
    reply("hyprspace-test:ping", "Invalid dispatcher")
    s.check("hyprspace then B unload leaves no handler pointing into either library")

    load(s.plugin)
    load(fixture)
    for value in ("first", "second"):
        s.ctl("dispatch", "hyprspace-test:replace", value)
        reply("submap", "fixture:" + value)
    unload(fixture)
    unmapped(fixture)
    unload(s.plugin)
    reply("submap", "Invalid dispatcher")
    s.check("same-type handler replacement while both plugins are loaded cannot restore a removed registration")

    # Also cover a native name already replaced before hyprspace installs.
    load(fixture)
    s.ctl("dispatch", "hyprspace-test:replace", "early")
    load(s.plugin)
    reply("submap", "fixture:early")
    unload(fixture)
    unmapped(fixture)
    unload(s.plugin)
    reply("submap", "Invalid dispatcher")
    s.check("a foreign replacement present at install is never retained across provider unload")

    # submap was deliberately removed by B; check a native dispatcher that
    # was only wrapped by hyprspace, then leave the plugin ready for other suites.
    s.ctl("dispatch", "workspace", "11")
    assert s.data("activeworkspace")["id"] == 11
    s.check("untouched native dispatchers retain their behavior after unloading")
    load(s.plugin)

File: test/integration/dispatchers.cpp
#include "generation.hpp"

// Independent library: unloading it must invalidate its dispatcher code.
#include <hyprland/src/plugins/PluginAPI.hpp>
#include <hyprland/src/managers/KeybindManager.hpp>

#include <stdexcept>

namespace {
    HANDLE handle;

    SDispatchResult raw(std::string) {
        return {.success = false, .error = "fixture:raw"};
    }

    void replace(const std::string& name, const std::string& value) {
        HyprlandAPI::addDispatcherV2(handle, name, [value](std::string) { return SDispatchResult{.success = false, .error = "fixture:" + value}; });
    }
} // namespace

APICALL EXPORT std::string PLUGIN_API_VERSION() {
    return HYPRLAND_API_VERSION;
}

APICALL EXPORT PLUGIN_DESCRIPTION_INFO PLUGIN_INIT(HANDLE plugin) {
    if (std::string{__hyprland_api_get_hash()} != __hyprland_api_get_client_hash())
        throw std::runtime_error("dispatcher test fixture ABI mismatch");
    handle = plugin;
    replace("hyprspace-test:ping", "original");
    HyprlandAPI::addDispatcher(handle, "hyprspace-test:legacy", [](std::string) {});
    g_pKeybindManager->m_dispatchers["hyprspace-test:raw"]     = raw;
    g_pKeybindManager->m_dispatchers["hyprspace-test:closure"] = [](std::string) { return raw({}); };
    HyprlandAPI::addDispatcherV2(handle, "hyprspace-test:replace", [](std::string value) {
        replace("submap", value);
        return SDispatchResult{};
    });
    return {"hyprspace-dispatcher-test", "Private dispatcher lifetime fixture", "hyprspace", "1"};
}

APICALL EXPORT void PLUGIN_EXIT() {
    g_pKeybindManager->m_dispatchers.erase("hyprspace-test:raw");
    g_pKeybindManager->m_dispatchers.erase("hyprspace-test:closure");
}
